flocker/ca/_ca.py

```diff
@@ -0,0 +1,225 @@
+# Copyright Hybrid Logic Ltd.  See LICENSE file for details.
+
+"""
+Low-level logic for a certificate authority.
+
+Uses ECDSA + SHA256? Or if that turns out to be too difficult RSA 4096-bit
++ SHA 256.
+"""
+
+import os
+
+from OpenSSL import crypto
+from pyrsistent import PRecord, field
+from twisted.internet.ssl import DistinguishedName, KeyPair, Certificate
+
+
+EXPIRY_20_YEARS = 60 * 60 * 24 * 365 * 20
+
+
+class CertificateAlreadyExistsError(Exception):
+    """
+    Error raised when a certificate file already exists.
+    """
+
+
+class KeyAlreadyExistsError(Exception):
+    """
+    Error raised when a keypair file already exists.
+    """
+
+
+class PathError(Exception):
+    """
+    Error raised when the directory for certificate files does not exist.
+    """
+
+
+class FlockerKeyPair(object):
+    """
+    KeyPair with added functionality for comparison and signing a request
+    object with additional extensions for generating a self-signed CA.
+    """
+    def __init__(self, keypair):
+        self.keypair = keypair
+
+    def __eq__(self, other):
+        if isinstance(other, FlockerKeyPair):
+            return self.keypair.dump() == other.keypair.dump()
+        return False
+
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+    def selfSignedCACertificate(self, dn, request, serial, expiry, digest):
```

Comment:
Judging by this camel case method name, I guess we do intend to upstream this to Twisted, so definitely add a ticket link and a comment about that.